testapi.pm

```diff
@@ -1252,25 +1252,30 @@ sub type_string {
         return;
     }
 
-    my $max_interval = $args{max_interval}       // 250;
-    my $wait         = $args{wait_screen_change} // 0;
-    my $wait_still   = $args{wait_still_screen}  // 0;
-    bmwqemu::log_call(string => $log, max_interval => $max_interval, wait_screen_changes => $wait, wait_still_screen => $wait_still);
+    my $max_interval   = $args{max_interval}       // 250;
+    my $wait           = $args{wait_screen_change} // 0;
+    my $wait_still     = $args{wait_still_screen}  // 0;
+    my $wait_timeout   = $args{timeout}            // 30;
+    my $wait_sim_level = $args{similarity_level}   // 47;
+    bmwqemu::log_call(string => $log, max_interval => $max_interval, wait_screen_changes => $wait, wait_still_screen => $wait_still,
+        timeout => $wait_timeout, similarity_level => $wait_sim_level);
     if ($wait) {
         # split string into an array of pieces of specified size
         # https://stackoverflow.com/questions/372370
         my @pieces = unpack("(a${wait})*", $string);
         for my $piece (@pieces) {
             wait_screen_change { query_isotovideo('backend_type_string', {text => $piece, max_interval => $max_interval}); };
-            if ($wait_still) {
-                wait_still_screen($wait_still);
+            if ($wait_still && !wait_still_screen(stilltime => $wait_still,
+                    timeout => $wait_timeout, similarity_level => $wait_sim_level)) {
+                die "wait_still_screen timed out after ${wait_timeout}s!!!";
             }
         }
     }
     else {
         query_isotovideo('backend_type_string', {text => $string, max_interval => $max_interval});
-        if ($wait_still) {
-            wait_still_screen($wait_still);
+        if ($wait_still && !wait_still_screen(stilltime => $wait_still,
+                timeout => $wait_timeout, similarity_level => $wait_sim_level)) {
+            die "wait_still_screen timed out after ${wait_timeout}s!!!";
```

Comment:
It seems we can extract this block from the if/else as it is called in both places. This would prevent the additional duplication. However I recommend you extend the unit tests first, then you can extract that part and see if tests still work or if I was wrong :)

Reply:
yup, that is my plan for today. Thanks for hints